### PR TITLE
storage/engine: allow caching of the resume key

### PR DESCRIPTION
**Pull request by petermattis**
Previously, the resume key was using the same underlying memory buffer
as the result data which meant that any caching of the resume key would
pin all of the result data in memory. Now we take care to make a copy of
the resume key. This addressess part of an egregious memory issue where
the `TxnCoordSender` was caching resume keys (refresh keys). Still to
figure out is why that cache isn't being freed when the txn finishes (or
even when the session ends).

Release note: None

### CODE REVIEW

**Comment by spencerkimball** [comment]
LGTM. I just wrote the same damn change. :-(

By the way, it's pretty clear that the SQL session holds onto the last implicit txn, so you might edit your PR description.

**Review by spencerkimball — APPROVED** [review]

**Comment by jordanlewis** [comment]
LGTM as far as fixing the leak, but like I said offline, it seems like it should be the responsibility of the code that intends to hold on to this memory for longer than expected to make the copy.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/engine/mvcc.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-L4NbAQDkZ5_fwefTaGg:-L4NbAQDkZ5_fwefTaGh:b-e76boc) ([raw file](https://github.com/cockroachdb/cockroach/blob/16f04a1bdae8b8977ec58397bcfa9f9903bec44d/pkg/storage/engine/mvcc.go#L1656)):*
> ```Go
> 	if resumeKey != nil {
> 		if reverse {
> 			resumeSpan = &roachpb.Span{Key: key, EndKey: resumeKey.Next()}
> ```

This will cause an unconditional extra copy if the scan is a reverse scan. Perhaps the copy should be done in this stanza, when `resumeKey != nil`, instead of in `buildScanResults`. That way you'd also only have to do the copy in one spot in the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-:-L4NayWJaYg0t_cUiVGp:bi58n03)*

**Comment by jordanlewis** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/engine/mvcc.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-L4NbAQDkZ5_fwefTaGg:-L4Nc1PSfSk7k6BQoHAJ:b-oevymq) ([raw file](https://github.com/cockroachdb/cockroach/blob/16f04a1bdae8b8977ec58397bcfa9f9903bec44d/pkg/storage/engine/mvcc.go#L1656)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This will cause an unconditional extra copy if the scan is a reverse scan. Perhaps the copy should be done in this stanza, when `resumeKey != nil`, instead of in `buildScanResults`. That way you'd also only have to do the copy in one spot in the code.
</blockquote></details>

Err, rather when `resumeKey != nil && !reverse`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/engine/mvcc.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-L4NbAQDkZ5_fwefTaGg:-L4Ne31XIF04ZkAEtAJ6:b-1athm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/16f04a1bdae8b8977ec58397bcfa9f9903bec44d/pkg/storage/engine/mvcc.go#L1656)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Err, rather when `resumeKey != nil && !reverse`
</blockquote></details>

Yeah, it's probably better to do it here. Changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343)*

**Comment by petermattis** [comment]
I'm not sure what is holding on to this memory. Clearly it has to do with the resume span, but precisely how is still a mystery. If either of you understands it, please enlighten me.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-:-L4NemWzIA1R4Jbk0Lwl:bco0lgs)*

**Comment by jordanlewis** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-:-L4NgEAIjxCoD4WsLoI0:bnfp4nl)*

**Comment by petermattis** [comment]
For posterity: Ah, I think I figured part (or all) of this out. `TxnMeta.RefreshReads` is holding on to the memory at that is causing the Go heap to grow very large. But once the query ends, there is nothing that triggers the GC to run to reclaim the memory. The Go GC is triggered by heap activity and while there is activity going on, the heap has grown to such a size that the background activity takes minutes for it to accumulate enough garbage for the GC to run. We can see this by running with `GODEBUG=gctrace=1`. The `^gc` lines show:

```
gc 9 @24.163s 0%: 0.015+1.7+0.030 ms clock, 0.12+0/2.9/7.0+0.24 ms cpu, 134->134->101 MB, 153 MB goal, 8 P
gc 10 @24.458s 0%: 0.009+1.9+0.056 ms clock, 0.076+0/2.9/7.2+0.44 ms cpu, 192->193->175 MB, 203 MB goal, 8 P
gc 11 @24.936s 0%: 0.011+2.3+0.071 ms clock, 0.095+0/2.5/7.7+0.57 ms cpu, 337->337->303 MB, 350 MB goal, 8 P
gc 12 @25.806s 0%: 0.013+4.0+0.035 ms clock, 0.10+0/4.0/11+0.28 ms cpu, 595->595->524 MB, 607 MB goal, 8 P
gc 13 @27.279s 0%: 0.012+2.9+0.078 ms clock, 0.097+0.12/4.7/7.2+0.62 ms cpu, 1024->1025->918 MB, 1048 MB goal, 8 P
gc 14 @29.883s 0%: 0.022+3.3+0.044 ms clock, 0.18+0/4.4/10+0.35 ms cpu, 1791->1792->1600 MB, 1837 MB goal, 8 P
gc 15 @35.078s 0%: 0.015+7.9+0.094 ms clock, 0.12+0/5.0/15+0.75 ms cpu, 3122->3123->2775 MB, 3200 MB goal, 8 P
gc 16 @155.120s 0%: 0.044+3.5+0.038 ms clock, 0.35+0.39/7.0/18+0.30 ms cpu, 3431->3431->75 MB, 5550 MB goal, 8 P
```

I started the query right around the `@24s` mark (`gc 9`). The query took ~12s to process during which the heap size grew to ~3GB. When the query finished, the GC “goal” (i.e. the trigger for the next GC) was set to 3.2GB of allocations. When the GC did eventually run, it collected a huge amount of memory.

The TLDR is I think the resume span fix is sufficient.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22343#-:-L4OT8I9b-huklFheWaX:b-8rw3o1)*
